File: substrate/research_workstation_interrogation_loop_compose.py
"""Research workstation interrogation loop compose (pure).

live_dispatched, pack_dispatched, record_persisted, prompts_injected,
live_router_authorized always False.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.research_interrogation_subagent_chase_compose import (
    ResearchInterrogationSubagentChaseCompose,
    ResearchInterrogationSubagentChaseComposeError,
    compose_research_interrogation_subagent_chase,
)
from substrate.workstation_record_prompt_model_decision_compose import (
    WorkstationRecordPromptModelDecisionCompose,
    WorkstationRecordPromptModelDecisionComposeError,
    compose_workstation_record_prompt_model_decision,
)


class ResearchWorkstationInterrogationLoopComposeError(ValueError):
    """Fail-closed validation for research workstation interrogation loop."""


@dataclass(frozen=True)
class ResearchWorkstationInterrogationLoopCompose:
    session_id: str
    parent_asset_id: str
    chase: ResearchInterrogationSubagentChaseCompose
    prompt_pack: WorkstationRecordPromptModelDecisionCompose
    loop_ready: bool
    live_dispatched: bool
    pack_dispatched: bool
    record_persisted: bool
    prompts_injected: bool
    live_router_authorized: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "chase": self.chase.to_dict(),
            "prompt_pack": self.prompt_pack.to_dict(),
            "loop_ready": self.loop_ready,
            "live_dispatched": False,
            "pack_dispatched": False,
            "record_persisted": False,
            "prompts_injected": False,
            "live_router_authorized": False,
            "notes": list(self.notes),
            "authority": (
                "research_workstation_interrogation_loop_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ResearchWorkstationInterrogationLoopComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def _build_loop_records(
    questions: list[object],
    prior: object | None,
) -> list[dict[str, Any]]:
    records: list[dict[str, Any]] = []
    if prior is not None:
        if not isinstance(prior, list):
            raise ResearchWorkstationInterrogationLoopComposeError(
                "prior_records must be an array when set"
            )
        for r in prior:
            if isinstance(r, dict):
                records.append(dict(r))
    for q in questions:
        if not isinstance(q, dict):
            raise ResearchWorkstationInterrogationLoopComposeError(
                "questions entries must be objects"
            )
        qid = _require_nonempty(q.get("question_id"), field="question_id")
        body = _require_nonempty(q.get("body"), field="body")
        records.append(
            {
                "record_id": f"q-{qid}",
                "kind": "question",
                "body": body,
                "source_ref": qid,
            }
        )
    if len(records) == 0:
        raise ResearchWorkstationInterrogationLoopComposeError(
            "loop requires ≥1 question or prior record"
        )
    return records


def compose_research_workstation_interrogation_loop(
    *,
    session_id: object,
    parent_asset_id: object,
    questions: object,
    chase_mode: object,
    user_prompt: object,
    selected_model_id: object,
    models: object,
    daily_cap_usd: object,
    spent_usd: object,
    operator_ack: object,
    prior_records: object | None = None,
    projected_cost_usd_high: object | None = None,
    projected_cost_usd_low: object | None = None,
    would_exceed: object | None = None,
    operator_override: object | None = None,
    source_families: object | None = None,
    bench_bests: object | None = None,
    focus_task: object | None = None,
    nd_shadow: object | None = None,
    mark_for_twin_record: object | None = None,
) -> ResearchWorkstationInterrogationLoopCompose:
    """Interrogate → chase → records → prompt pack. Never dispatches/injects."""
    if not isinstance(operator_ack, bool):
        raise ResearchWorkstationInterrogationLoopComposeError(
            "operator_ack must be an explicit boolean"
        )
    session = _require_nonempty(session_id, field="session_id")
    parent = _require_nonempty(parent_asset_id, field="parent_asset_id")
    if not isinstance(questions, list) or len(questions) == 0:
        raise ResearchWorkstationInterrogationLoopComposeError(
            "questions must be a non-empty array"
        )

    notes: list[str] = [
        "live_dispatched=false — chase slots are pure intent",
        "pack_dispatched=false",
        "record_persisted=false — session substrate advisory only",
        "prompts_injected=false — proposed prompt envelope only",
        "live_router_authorized=false — operator selects model",
    ]

    mark_twin = True if mark_for_twin_record is None else mark_for_twin_record
    if not isinstance(mark_twin, bool):
        raise ResearchWorkstationInterrogationLoopComposeError(
            "mark_for_twin_record must be boolean when set"
        )

    try:
        chase = compose_research_interrogation_subagent_chase(
            session_id=session,
            parent_asset_id=parent,
            questions=questions,
            chase_mode=chase_mode,
            would_exceed=would_exceed,
            operator_ack=operator_ack,
            operator_override=operator_override,
            selected_model_id=selected_model_id,
            source_families=source_families,
            mark_for_twin_record=mark_twin,
        )
    except ResearchInterrogationSubagentChaseComposeError as e:
        raise ResearchWorkstationInterrogationLoopComposeError(str(e)) from e
    notes.extend(f"[chase] {n}" for n in chase.notes)

    records = _build_loop_records(questions, prior_records)
    notes.append(
        f"session_records={len(records)} from questions+prior (caller-supplied only)"
    )

    try:
        prompt_pack = compose_workstation_record_prompt_model_decision(
            session_id=session,
            parent_asset_id=parent,
            records=records,
            user_prompt=user_prompt,
            selected_model_id=selected_model_id,
            models=models,
            daily_cap_usd=daily_cap_usd,
            spent_usd=spent_usd,
            operator_ack=operator_ack,
            projected_cost_usd_high=projected_cost_usd_high,
            projected_cost_usd_low=projected_cost_usd_low,
            bench_bests=bench_bests,
            focus_task=(
                "deep_research" if focus_task is None else focus_task
            ),
            nd_shadow=nd_shadow,
        )
    except WorkstationRecordPromptModelDecisionComposeError as e:
        raise ResearchWorkstationInterrogationLoopComposeError(str(e)) from e
    notes.extend(f"[prompt] {n}" for n in prompt_pack.notes)

    loop_ready = (
        chase.chase_ready is True
        and prompt_pack.pack_ready is True
        and operator_ack is True
    )
    if loop_ready:
        notes.append(
            "loop_ready=true — interrogate→chase→record→prompt pack ready; still pure"
        )
    else:
        notes.append(
            "loop_ready=false — chase, prompt pack, or operator_ack gate open"
        )

    if (
        chase.live_dispatched is not False
        or chase.pack_dispatched is not False
        or chase.record_persisted is not False
        or chase.prompts_injected is not False
        or prompt_pack.record_persisted is not False
        or prompt_pack.prompts_injected is not False
        or prompt_pack.live_router_authorized is not False
    ):
        raise ResearchWorkstationInterrogationLoopComposeError(
            "invariant: honesty flags must remain false"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "record_persisted=false",
            "prompts_injected=false",
            "live_router_authorized=false",
        )
    )

    return ResearchWorkstationInterrogationLoopCompose(
        session_id=session,
        parent_asset_id=parent,
        chase=chase,
        prompt_pack=prompt_pack,
        loop_ready=loop_ready,
        live_dispatched=False,
        pack_dispatched=False,
        record_persisted=False,
        prompts_injected=False,
        live_router_authorized=False,
        notes=tuple(notes),
        authority="research_workstation_interrogation_loop_compose_advisory",
    )


def format_research_workstation_interrogation_loop_summary(
    c: ResearchWorkstationInterrogationLoopCompose,
) -> str:
    return (
        f"loop_ready={c.loop_ready} · chase_slots={c.chase.slot_count} · "
        f"would_exceed={c.prompt_pack.would_exceed} · "
        f"live_dispatched=false · record_persisted=false · prompts_injected=false"
    )


__all__ = [
    "ResearchWorkstationInterrogationLoopCompose",
    "ResearchWorkstationInterrogationLoopComposeError",
    "compose_research_workstation_interrogation_loop",
    "format_research_workstation_interrogation_loop_summary",
]

File: apps/reading/src/api/researchWorkstationInterrogationLoopCompose.ts
/**
 * Research workstation interrogation loop compose (pure).
 *
 * Operator vision: live in the research workstation — interrogate material,
 * send subagents to chase questions, record insights/questions recursively,
 * and feed that substrate into the next prompt + model decision (with budget
 * projection). Pure intent only — never live-dispatches or injects prompts.
 *
 * live_dispatched always false.
 * pack_dispatched always false.
 * record_persisted always false.
 * prompts_injected always false.
 * live_router_authorized always false.
 */

import {
  composeResearchInterrogationSubagentChase,
  type ChaseMode,
  type ChaseQuestion,
  type ResearchInterrogationSubagentChaseCompose,
  type SourceFamilyHint,
} from "./researchInterrogationSubagentChaseCompose";
import {
  composeWorkstationRecordPromptModelDecision,
  type WorkstationRecordPromptModelDecisionCompose,
} from "./workstationRecordPromptModelDecisionCompose";
import type { SessionRecordItem } from "./workstationSessionInsightRecordCompose";
import type { ModelOption } from "./modelDecisionPromptCompose";
import type {
  BenchTaskBest,
  NotDiamondShadowRec,
} from "./settingsModelDriverTabCompose";

export interface ResearchWorkstationInterrogationLoopInput {
  session_id: string;
  parent_asset_id: string;
  /** Questions operator is wrestling with — caller-supplied only. */
  questions: ChaseQuestion[];
  chase_mode: ChaseMode;
  /** Optional prior insights/claims (caller-supplied). */
  prior_records?: SessionRecordItem[] | null;
  user_prompt: string;
  selected_model_id: string;
  models: ModelOption[];
  daily_cap_usd: number | null;
  spent_usd: number | null;
  projected_cost_usd_high?: number | null;
  projected_cost_usd_low?: number | null;
  would_exceed?: boolean | null;
  operator_override?: boolean;
  source_families?: SourceFamilyHint[] | null;
  bench_bests?: BenchTaskBest[] | null;
  focus_task?: string | null;
  nd_shadow?: NotDiamondShadowRec | null;
  operator_ack: boolean;
  /** Default true — mark chase + records for twin/prompt substrate. */
  mark_for_twin_record?: boolean;
  mark_for_prompt_context?: boolean;
}

export interface ResearchWorkstationInterrogationLoopCompose {
  session_id: string;
  parent_asset_id: string;
  chase: ResearchInterrogationSubagentChaseCompose;
  prompt_pack: WorkstationRecordPromptModelDecisionCompose;
  /**
   * True when chase.chase_ready and prompt_pack.pack_ready.
   * Still never dispatches or injects.
   */
  loop_ready: boolean;
  live_dispatched: false;
  pack_dispatched: false;
  record_persisted: false;
  prompts_injected: false;
  live_router_authorized: false;
  notes: string[];
  authority: "research_workstation_interrogation_loop_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Build session records from chase questions + prior records.
 * Never invents insight bodies beyond caller-supplied question text.
 */
function buildLoopRecords(
  questions: ChaseQuestion[],
  prior: SessionRecordItem[] | null | undefined,
): SessionRecordItem[] {
  const records: SessionRecordItem[] = [];
  if (prior != null) {
    if (!Array.isArray(prior)) {
      throw new Error("prior_records must be an array when set");
    }
    for (const r of prior) {
      records.push(r);
    }
  }
  for (const q of questions) {
    records.push({
      record_id: `q-${q.question_id}`,
      kind: "question",
      body: q.body,
      source_ref: q.question_id,
    });
  }
  if (records.length === 0) {
    throw new Error("loop requires ≥1 question or prior record");
  }
  return records;
}

/**
 * Compose interrogation chase → session records → prompt/model decision.
 * Never live-dispatches chases; never persists; never injects prompts.
 */
export function composeResearchWorkstationInterrogationLoop(
  input: ResearchWorkstationInterrogationLoopInput,
): ResearchWorkstationInterrogationLoopCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  const session_id = requireNonEmpty(input.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    input.parent_asset_id,
    "parent_asset_id",
  );
  if (!Array.isArray(input.questions) || input.questions.length === 0) {
    throw new Error("questions must be a non-empty array");
  }

  const notes: string[] = [
    "live_dispatched=false — chase slots are pure intent",
    "pack_dispatched=false",
    "record_persisted=false — session substrate advisory only",
    "prompts_injected=false — proposed prompt envelope only",
    "live_router_authorized=false — operator selects model",
  ];

  const mark_for_twin =
    input.mark_for_twin_record === undefined
      ? true
      : input.mark_for_twin_record;
  if (typeof mark_for_twin !== "boolean") {
    throw new Error("mark_for_twin_record must be boolean when set");
  }

  // Budget honesty for chase: prefer explicit would_exceed; else null
  const would_exceed =
    input.would_exceed === undefined ? null : input.would_exceed;
  if (
    would_exceed !== null &&
    would_exceed !== undefined &&
    typeof would_exceed !== "boolean"
  ) {
    throw new Error("would_exceed must be boolean or null");
  }

  const chase = composeResearchInterrogationSubagentChase({
    session_id,
    parent_asset_id,
    questions: input.questions,
    chase_mode: input.chase_mode,
    would_exceed,
    operator_override: input.operator_override,
    selected_model_id: input.selected_model_id,
    source_families: input.source_families,
    operator_ack: input.operator_ack,
    mark_for_twin_record: mark_for_twin,
  });
  notes.push(...chase.notes.map((n) => `[chase] ${n}`));

  const records = buildLoopRecords(input.questions, input.prior_records);
  notes.push(
    `session_records=${records.length} from questions+prior (caller-supplied only)`,
  );

  const prompt_pack = composeWorkstationRecordPromptModelDecision({
    session_id,
    parent_asset_id,
    records,
    user_prompt: input.user_prompt,
    selected_model_id: input.selected_model_id,
    models: input.models,
    daily_cap_usd: input.daily_cap_usd,
    spent_usd: input.spent_usd,
    projected_cost_usd_high: input.projected_cost_usd_high,
    projected_cost_usd_low: input.projected_cost_usd_low,
    bench_bests: input.bench_bests,
    focus_task: input.focus_task ?? "deep_research",
    nd_shadow: input.nd_shadow,
    operator_ack: input.operator_ack,
  });
  notes.push(...prompt_pack.notes.map((n) => `[prompt] ${n}`));

  const loop_ready =
    chase.chase_ready === true &&
    prompt_pack.pack_ready === true &&
    input.operator_ack === true;

  if (loop_ready) {
    notes.push(
      "loop_ready=true — interrogate→chase→record→prompt pack ready; still pure",
    );
  } else {
    notes.push(
      "loop_ready=false — chase, prompt pack, or operator_ack gate open",
    );
  }

  if (
    chase.live_dispatched !== false ||
    chase.pack_dispatched !== false ||
    chase.record_persisted !== false ||
    chase.prompts_injected !== false ||
    prompt_pack.record_persisted !== false ||
    prompt_pack.prompts_injected !== false ||
    prompt_pack.live_router_authorized !== false
  ) {
    throw new Error("invariant: honesty flags must remain false");
  }

  notes.push("live_dispatched=false");
  notes.push("record_persisted=false");
  notes.push("prompts_injected=false");
  notes.push("live_router_authorized=false");

  return {
    session_id,
    parent_asset_id,
    chase,
    prompt_pack,
    loop_ready,
    live_dispatched: false,
    pack_dispatched: false,
    record_persisted: false,
    prompts_injected: false,
    live_router_authorized: false,
    notes,
    authority: "research_workstation_interrogation_loop_compose_advisory",
  };
}

export function formatResearchWorkstationInterrogationLoopSummary(
  c: ResearchWorkstationInterrogationLoopCompose,
): string {
  return (
    `loop_ready=${c.loop_ready} · chase_slots=${c.chase.slot_count} · ` +
    `would_exceed=${c.prompt_pack.would_exceed} · ` +
    `live_dispatched=false · record_persisted=false · prompts_injected=false`
  );
}
